### DISCUSSION TITLE
I added several functions to increase the CPU usage of my project from 0-2% to 0-1% under Window.

### DISCUSSION THREAD

**Discussion by xuyun018** [fltk/fltk#1036] [Development]
Fl_Widget.h
```c++
 bool needs_keyboard() const {
    return (flags_ & NEEDS_KEYBOARD);
  }
```
Change to
```c++
 bool needs_keyboard() const {
    return ((flags_ & NEEDS_KEYBOARD) != 0);
  }
```

In Windows:
Fl_GDI_Graphics_Driver_image.cxx

Add the following three functions before the function 'innards'.

[Fl_GDI_Graphics_Driver_image.txt](https://github.com/user-attachments/files/16503796/Fl_GDI_Graphics_Driver_image.txt)

//---------------------------------------------------------
Add virtual functions to the class' Fl_Graphics_Driver.h '
```c++
virtual void draw_xyimage(const uchar* buf, int X, int Y, int W, int H);
```
```c++
virtual void draw_xyimage_unscaled(const uchar* buf, int X, int Y, int W, int H);
```

class Fl_Scalable_Graphics_Driver:
```c++
void draw_xyimage(const uchar* buf, int X, int Y, int W, int H) FL_OVERRIDE;
```

//---------------------------------------------------------
Add virtual functions to the class' Fl_Graphics_Driver.cxx '
```c++
void Fl_Graphics_Driver::draw_xyimage(const uchar* buf, int X, int Y, int W, int H) {}

class Fl_Scalable_Graphics_Driver:
void Fl_Scalable_Graphics_Driver::draw_xyimage_unscaled(const uchar* buf, int X, int Y, int W, int H) {}

void Fl_Scalable_Graphics_Driver::draw_xyimage(const uchar* buf, int X, int Y, int W, int H) {
	draw_xyimage_unscaled(buf, X, Y, W, H);
}
```
//---------------------------------------------------------
Add functions to the class' Fl_GDI_Graphics_Driver.h '
```c++
void draw_xyimage_unscaled(const uchar* buf, int X, int Y, int W, int H) FL_OVERRIDE;
```

//---------------------------------------------------------
Add functions to the class' Fl_GDI_Graphics_Driver_image.cxx '
```c++
void Fl_GDI_Graphics_Driver::draw_xyimage_unscaled(const uchar* buf, int x, int y, int w, int h){
	xyinnards(buf, x, y, w, h, 3, 0, 0, gc_);
}
```

//---------------------------------------------------------
Add functions to the class' fl_draw.h '
```c++
unsigned int fl_create_xyimage(uchar *dst, unsigned int dst_size, unsigned int *_row_size,
	const uchar *buf, int _w, int _h,
	int delta, int linedelta, int depth);
void fl_xyimage_pixel(uchar *buf, int _w, int _h, int row_size, int _x, int _y,
	unsigned char _r, unsigned char _g, unsigned char _b);

inline void fl_draw_xyimage(const uchar *buf, int X, int Y, int W, int H) {
	fl_graphics_driver->draw_xyimage(buf, X, Y, W, H);
}
```

**Comment by xuyun018** [comment]
In Ubuntu:

Fl_Xlib_Graphics_Driver_image.cxx

[Fl_Xlib_Graphics_Driver_image.txt](https://github.com/user-attachments/files/16503804/Fl_Xlib_Graphics_Driver_image.txt)

**Comment by xuyun018** [comment]
Fl_Graphics_Driver.h

class Fl_Graphics_Driver:
```c++
friend void fl_draw_xyimage(const uchar* buf, int X, int Y, int W, int H);
```

The unit modified under Windows is the "Fl_GDI_Craphics_Driver" unit, while under UBuntu it is the "Fl_Xlib_Graphics_Driver" unit

**Comment by erco77** [comment]
OP, when posting code inline, please enclose the code with triple backquotes to preserve indent. To enable syntax highlighting, include the "c++" modifier. Example:

    ```c++
    Your code goes here
    ```
If you do that literally in your comments, the code will have its indenting preserved.

Also, please use one type of indent for if/for/etc. For suggesting additions to FLTK, it's best to follow our [coding standards](https://www.fltk.org/cmp.php) which uses 2 space indent, and k+r style indent, e.g.
```c++
void some_function() {
  if (something) {
    do_something_else();
  }
  for (..) {
    stuff;
  }
}
```

**Reply by xuyun018** [reply]
I am a beginner and accustomed to using VS IDE. I will pay attention next time.

**Comment by xuyun018** [comment]
Testing project in windows.
[Uploading fltk_win_test.zip…]()

**Reply by erco77** [reply]
Please note you can edit your existing comments to make fixes/changes by clicking on the triple horizontal dots at the right of each comment. Example:

![image](https://github.com/user-attachments/assets/4e997d4d-712c-482e-89ea-ff3bcbfe9213)

Try to keep inline code snippets in comments no longer than 1/2 a page. For longer code, you can attach single c++ files by making a copy on your machine ending in .txt, and attach that. This way users reading your messages can just click the .txt file and see it, without needing to unzip files. Always try to reduce your code examples to single files and remove as much unrelated information as possible. Anything too complicated will decrease the likelihood of getting timely responses.

**Reply by xuyun018** [reply]
OK,Tks

**Comment by xuyun018** [comment]
I noticed that the original 'innards' function (including GDI and Xlib) may require memory allocation and will definitely perform data conversion. When the image size is large and frequently needs to be drawn, the performance will be low.
And when it comes to true color images, the size of the original image data is similar to the size of the internal image data for display, so I wrote a function to directly manipulate the internal image data, which can avoid data conversion every time I draw to improve speed.
